pkg/linter/lintOverlappingSubnets.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2023- IBM Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package linter
+
+import (
+	"fmt"
+
+	"github.com/np-guard/models/pkg/ipblock"
+	"github.com/np-guard/vpc-network-config-analyzer/pkg/vpcmodel"
+)
+
+const overlappingSubnetsName = "overlapping-subnets"
+
+// overlapSubnets: overlapping subnet ranges (relevant mostly for the multiple VPCs use case)
+type overlappingSubnetsLint struct {
+	basicLinter
+}
+
+// a couple of overlapping subnets
+type overlapSubnets struct {
+	overlapSubnets  [2]vpcmodel.Subnet
+	overlapIPBlocks *ipblock.IPBlock
+}
+
+// /////////////////////////////////////////////////////////
+// lint interface implementation for overlapSubnets
+// ////////////////////////////////////////////////////////
+func (lint *overlappingSubnetsLint) lintName() string {
+	return overlappingSubnetsName
+}
+
+func (lint *overlappingSubnetsLint) lintDescription() string {
+	return "Overlapping CIDR ranges between different subnets"
+}
+
+func (lint *overlappingSubnetsLint) check() error {
+	allSubnets := []vpcmodel.Subnet{}
+	for _, config := range lint.configs {
+		if config.IsMultipleVPCsConfig {
+			continue
+		}
+		allSubnets = append(allSubnets, config.Subnets...)
+	}
+	for i, subnet1 := range allSubnets {
+		subnet1IPBlock, err1 := ipblock.FromCidr(subnet1.CIDR())
+		if err1 != nil {
+			return err1
+		}
+		for _, subnet2 := range allSubnets[i+1:] {
+			subnet2IPBlock, err2 := ipblock.FromCidr(subnet2.CIDR())
+			if err2 != nil {
+				return err2
+			}
+			intersectIPBlock := subnet1IPBlock.Intersect(subnet2IPBlock)
+			if !intersectIPBlock.IsEmpty() {
+				if subnetStr(subnet1) > subnetStr(subnet2) {
+					subnet1, subnet2 = subnet2, subnet1
+				}
+				lint.addFinding(&overlapSubnets{overlapSubnets: [2]vpcmodel.Subnet{subnet1, subnet2}, overlapIPBlocks: intersectIPBlock})
+			}
+		}
+	}
+	return nil
+}
+
+///////////////////////////////////////////////////////////
+// finding interface implementation for overlapSubnets
+//////////////////////////////////////////////////////////
+
+func (finding *overlapSubnets) vpc() []string {
+	return []string{finding.overlapSubnets[0].VPC().Name(), finding.overlapSubnets[1].VPC().Name()}
+}
+
+func (finding *overlapSubnets) string() string {
+	subnet1 := finding.overlapSubnets[0]
+	subnet2 := finding.overlapSubnets[1]
+	overlapStr := ""
+	if finding.overlapIPBlocks.String() == subnet1.CIDR() && subnet1.CIDR() == subnet2.CIDR() {
+		overlapStr = " overlap in the entire subnets' CIDR range"
+	} else {
+		overlapStr = fmt.Sprintf(" overlap in %s", finding.overlapIPBlocks.String())
+	}
```

Comment:
is it necessary to differentiate between these cases in the report?

Reply:
I think it is more readable this way. If its practically the same subnet the information is relevant.